Simple implementation of lanczos scaling

simmplecoder opened the pull request:
### Description

Implementation of downscaling through resampling by Lanczos. It seems to perform reasonably fast and have reasonable quality of downscaling.

*Here, describe what does this pull request do?*

### Environment

Should run on all supported configurations 1.70+

### References

https://github.com/BoostGSoC19/gil-olzhas/projects/1

### Tasklist

- [X] Prototype
- [X] Refactoring into own header
<s>- [ ] Hide details into include/boost/gil/detail/</s> Unapplicable
- [X] Add test case(s)
- [ ] Ensure all CI builds pass
- [ ] Review and approve

simmplecoder committed:
Simple implementation of lanczos scaling

simmplecoder committed:
Simple implementation of lanczos scaling

stefanseefeld commented on test/image_processing/lanczos_scaling.cpp:
```diff
@@ -0,0 +1,82 @@
+namespace gil = boost::gil;
+namespace mpl = boost::mpl;
+using image_view = decltype(gil::view(std::declval<gil::rgb8_image_t&>()));
+double sinc(double x) {
```
coding style nit-pick: opening brackets should be written on their own line.

stefanseefeld commented on test/image_processing/lanczos_scaling.cpp:
```diff
@@ -0,0 +1,82 @@
+    return std::sin(x * pi) / (x * pi);
+}
+
+double lanczos(long int x, long int a) {
```
I assume `lanczos`, `lanczos_at`, and `scale_lanczos` are the three new functions you are proposing (and which the current file is supposed to be testing, right ? Shouldn't the functions themselves then be defined elsewhere, to be usable by `Boost.GIL` users ?

stefanseefeld replied:
In either case, these functions need (inline) documentation. As we are using doxygen, I'd suggest simple doxygen-style comments such as
```
//! bla bla bla
```
potentially with inline markdown markup. A simple description of what the function is doing, and, if necessary, what the input arguments and return value mean.

simmplecoder replied:
I believe commenting is the only thing left from the first round of review from you

stefanseefeld commented on test/image_processing/lanczos_scaling.cpp:
```diff
@@ -0,0 +1,82 @@
+//#include <boost/gil/image_processing/scaling.hpp>
```
I assume the above (still non-existent) header is supposed to contain the functions, eventually. What's the reason you haven't used that yet ?

mloskot commented on test/image_processing/lanczos_scaling.cpp:
```diff
@@ -0,0 +1,82 @@
+
+namespace gil = boost::gil;
+namespace mpl = boost::mpl;
+using image_view = decltype(gil::view(std::declval<gil::rgb8_image_t&>()));
```
Is this algorithm going to be rgb8 specific only?

stefanseefeld replied:
This may be enough for a quick prototype of an algorithm, but eventually all of the functionality needs to be as generic as possible, i.e. applicable to as wide a range of channel / pixel / image types as possible.

mloskot commented on test/image_processing/lanczos_scaling.cpp:
```diff
@@ -0,0 +1,82 @@
+#include <boost/gil/rgb.hpp>
+#include <boost/gil/pixel.hpp>
+
+#include <boost/mpl/int.hpp>
```
Please, do not use Boost.MPL. Instead, use C++11 and Boost.MP11

mloskot commented on test/image_processing/lanczos_scaling.cpp:
```diff
@@ -0,0 +1,82 @@
+    output_view(target_x, target_y) = gil::rgb8_pixel_t(result_red_channel, result_green_channel, result_blue_channel);
+}
+
+void scale_lanczos(image_view input_view, image_view output_view, long int a) {
```
If it is not going to be a template, I'd use `rgb8_image_t::view_t`. No need for cryptic type(def).

mloskot commented on test/image_processing/lanczos_scaling.cpp:
```diff
@@ -0,0 +1,82 @@
+    if (x == 0) {
+        return 1;
+    }
+    if (x > -a && x < a) {
```
`-a < x && x < a` reads clearer as range to me

simmplecoder replied:
Not sure if there is a consensus on this, so leaving this open for now. @stefanseefeld , what do you think?

stefanseefeld replied:
I agree with @mloskot.

mloskot commented on include/boost/gil/image_processing/scaling.hpp:
```diff
@@ -5,6 +5,29 @@
+/// \defgroup ScalingAlgorithms
+/// \brief Algorthims suitable for rescaling
+///
+/// These algorithms are used to improve image
```
@simmplecoder You may go for longer lines, up to ~~90~~ 100 (https://github.com/boostorg/gil/commit/5024f6ffdddecf4c0432892bf77fc5cf2ae1c1f8) characters :-)

simmplecoder replied:
@mloskot , thanks, helped on one occasion, but unfortunately there was a place where it still was too long. Resolving this.

mloskot commented on include/boost/gil/image_processing/numeric.hpp:
```diff
@@ -2,21 +2,40 @@
 
 namespace boost{
```
coding style nit-pick: `namespace boost { namespace gil {` is more widely used across GILL code

mloskot commented on include/boost/gil/image_processing/scaling.hpp:
```diff
@@ -5,6 +5,29 @@
+/// artifacts.
 template <typename ImageView>
 void lanczos_at(long int source_x, long int source_y, 
                 long int target_x, long int target_y, 
```
coding style nit-pick: Slightly different alignment has been agreed for GIL:

https://github.com/BoostGSoC19/gil-olzhas/blob/a123653127d6619c3bd211f74114d9962b5a4bf3/include/boost/gil/algorithm.hpp#L137-L144

GIL function names can be longer (as preferred descriptive than cryptic), so ligning to the right has this disadvantage:

```
void my_super_duper_algorith_better_than_lanczos_at(long int source_x, long int source_y, 
                                                    long int target_x, long int target_y, 
```

```
void my_super_duper_algorith_better_than_lanczos_at(long int source_x, 
                                                    long int source_y, 
                                                    long int target_x,
                                                    long int target_y, 
```

Having big blank space on the left is not as friendly for left-to-right reading.

simmplecoder replied:
Should I write one argument per line or concatenate if readable?

mloskot replied:
It's all driven by line length limit:
If all fit in single line up to 90 characters limit, then prefer single line.
If single line would be longer than 90, then split one argument per line as shown earlier.

stefanseefeld replied:
Actually (since we are talking about coding style): In Mateusz' example above, I'd prefer the first argument to follow the opening paren without newline, and the other arguments to be vertically aligned with the first. That's just a matter of readability: avoid newlines if not needed to make a statement (or expression) fit into a line.

mloskot replied:
@stefanseefeld I can't recall we've made a clear agreement on that, so there is no strict guideline. Or perhaps it was discussed as part of some of the many code reviews.
IMO, more important quality is to have style consistent within a file, at least.

I just would really like to avoid the big empty multiline gaps, that are e.g. followed by 2-3 lines of function definition. For example, this is too much of blank space as I believe most of left-to-right readers anchor their eyes at left, not where semi-colons are :-)

----

![image](https://user-images.githubusercontent.com/80741/58614227-1d60fe80-82b8-11e9-9181-ff346b09b7f8.png)

-----

![image](https://user-images.githubusercontent.com/80741/58614321-5ef1a980-82b8-11e9-8cff-a3569664131d.png)

-----

@stefanseefeld  I'd rather increase the line length limit to 100 or 110 characters. Shall we?

mloskot replied:
@simmplecoder You can go wider ;-)
https://github.com/boostorg/gil/commit/5024f6ffdddecf4c0432892bf77fc5cf2ae1c1f8

stefanseefeld commented on include/boost/gil/image_processing/scaling.hpp:
```diff
@@ -1 +1,65 @@
+template <typename ImageView>
+void lanczos_at(long int source_x, long int source_y, 
+                long int target_x, long int target_y, 
+                ImageView input_view, ImageView output_view, long int a) {
```
A separate coding style issue (which I raised previously): new scopes (such as function scopes) should be opened on a new line. Read: please put the opening bracket onto its own line.

stefanseefeld commented on test/image_processing/lanczos_scaling.cpp:
```diff
@@ -1,73 +1,16 @@
+    namespace gil = boost::gil;
     if (argc != 3) {
         std::cerr << "usage: " << argv[0] << " <input-filename.png> <output-filename.png>\n";
         throw std::invalid_argument("invalid cli arguments");
```
I don't think throwing an exception from `main()` is meaningful. There is no scope to catch this exception in, so you could just as well `return -1`, after printing a suitable error message to `std::cerr`.

simmplecoder replied:
I added real tests now, main is mostly empty. I will keep this in mind in the future.

simmplecoder committed:
Refactor lanczos into separate header

This commit moves the algorithm
into its own header. It also provides
templated interface now, utilizing
a little bit of decltype and declval
to resolve types for lambdas and stuff

simmplecoder committed:
Refactor lanczos into separate header

This commit moves the algorithm
into its own header. It also provides
templated interface now, utilizing
a little bit of decltype and declval
to resolve types for lambdas and stuff

simmplecoder committed:
Zero pixel at start of lanczos_at

This commit fixes a possible bug and
fixes some style incosistencies

simmplecoder committed:
Zero pixel at start of lanczos_at

This commit fixes a possible bug and
fixes some style incosistencies

simmplecoder committed:
Improve documentation for Lanczos

This commit adds documentation which
describes when to use the algorithm,
a brief description of how it is
supposed to work, and some caution
on the quality of the output

simmplecoder committed:
Improve documentation for Lanczos

This commit adds documentation which
describes when to use the algorithm,
a brief description of how it is
supposed to work, and some caution
on the quality of the output

simmplecoder committed:
Address style issues and fix warnings

This commit fixes a style issue realted
to namespace declaration and
fixes a warning in added zeroing of
a pixel at start of lanczos

simmplecoder committed:
Address style issues and fix warnings

This commit fixes a style issue realted
to namespace declaration and
fixes a warning in added zeroing of
a pixel at start of lanczos

simmplecoder committed:
text-realign function arguments

simmplecoder committed:
text-realign function arguments

simmplecoder committed:
fix formatting issues

simmplecoder committed:
fix formatting issues

simmplecoder committed:
Implement handful of sanity tests

This commit introduces a couple of
sanity tests such as black image
scaling to black image, and lanczos
response being 0 at x = 0

simmplecoder committed:
Implement handful of sanity tests

This commit introduces a couple of
sanity tests such as black image
scaling to black image, and lanczos
response being 0 at x = 0

simmplecoder committed:
bracket on newline for for loops

simmplecoder committed:
bracket on newline for for loops

simmplecoder committed:
add lanczos scaling to tests

simmplecoder committed:
add lanczos scaling to tests

simmplecoder committed:
more precision in lanczos calculation

This commit migrates integral values
to double precision and uses PI
provided by boost.Math. These changes
solve downscaled image being darker

simmplecoder committed:
more precision in lanczos calculation

This commit migrates integral values
to double precision and uses PI
provided by boost.Math. These changes
solve downscaled image being darker

simmplecoder committed:
Merge branch 'lanczos' of https://github.com/BoostGSoC19/gil-olzhas into lanczos

simmplecoder committed:
Add Jamfile for ip test directory

Add Jamfile for image_processing
test directory and build-project
from outer test directory

stefanseefeld commented:
It appears the `travis` build configuration explicitly lists the tests to run (https://github.com/BoostGSoC19/gil-olzhas/blob/develop/.ci/build-and-test.sh#L41-L44), so you need to add the new test subdirectory there, too.

mloskot commented:
@stefanseefeld Since the `image_processing` is in core part i.e. `include/boost/gil/image_processing/`, shouldn't the test directory go into ` test/core/image_processing/`? What do you think?

If yes, then no need to list the directory in the CI scripts explicitly.

stefanseefeld commented:
@mloskot I have to admit I haven't quite understood what "core" actually stands for. Anything that's not an "extension" ?
Arguably, we could move everything under `test/core/` directly into `test/`. Let's keep that a separate discussion, though.

stefanseefeld reviewed (approved):
This looks good to me now. Thanks !

mloskot commented:
@stefanseefeld 
> I have to admit I haven't quite understood what "core" actually stands for. Anything that's not an "extension" ?

Yes, `core` is a synonym to core features i.e. w/o extensions, w/o test utilities, w/o benchmarks, etc.

> Arguably, we could move everything under test/core/ directly into test/.

There are good reasons why I kept it in the subdirectory sibling to otehrs, not parent.

> Let's keep that a separate discussion, though.

Sure, e.g. boost-gil

mloskot commented on .ci/build-and-test.sh:
```diff
@@ -42,3 +42,4 @@ set -euv
 ./b2 -j $JOBS $B2_OPTIONS toolset=$TOOLSET variant=$VARIANT cxxstd=$CXXSTD libs/gil/test/extension/numeric
 ./b2 -j $JOBS $B2_OPTIONS toolset=$TOOLSET variant=$VARIANT cxxstd=$CXXSTD libs/gil/test/extension/toolbox
 ./b2 -j $JOBS $B2_OPTIONS toolset=$TOOLSET variant=$VARIANT cxxstd=$CXXSTD libs/gil/test/extension/io//simple
+./b2 -j $JOBS $B2_OPTIONS toolset=$TOOLSET variant=$VARIANT cxxstd=$CXXSTD libs/gil/test/image_processing
```
@simmplecoder It is yet to be argued between me and @stefanseefeld  about the `test/image_processing` location in near future. 

Meanwhile, I'd suggest you to do this:

1. Move `test/image_processing` into `test/core/image_processing` (this is what Miral does too).
2. Remove this `b2` invocation from ` .ci/build-and-test.sh`
3. Enjoy the fact that your tests will be built automatically, *without* tweaking any CI configs.
4. Enjoy the fact your tests will be built *before* extensions, as they should be :)

simmplecoder committed:
Add IP test directory to ci build

This commit adds a line at the end
of .ci/build-and-test.sh to include
image processing tests in CI builds

simmplecoder committed:
Remove redundant lines from Jamfile

Simplify Jamfile at
test/image_processing

simmplecoder committed:
Rewrite range condition

x > -a && x < a exchange with
-a < x && x < a in lanczos

stefanseefeld commented on .ci/get-boost.sh:
```diff
@@ -42,6 +42,8 @@ git submodule --quiet update --init $GIT_SUBMODULE_OPTS \
     libs/filesystem \
     libs/integer \
     libs/iterator \
+    libs/lexical_cast \
```
We shouldn't depend on that.

stefanseefeld commented on include/boost/gil/detail/math.hpp:
```diff
@@ -0,0 +1,3 @@
+namespace boost{namespace gil{
+    static constexpr double pi = 3.14159265358979323846;
+}}
```
missing trailing newline

mloskot replied:
and spaces in `boost{namespace gil{`  😄

stefanseefeld commented on test/core/image_processing/Jamfile:
```diff
@@ -0,0 +1,3 @@
+import testing ;
+
+run lanczos_scaling.cpp ;
```
missing trailing newline.

simmplecoder committed:
Add newline at the end of files

stefanseefeld commented on test/CMakeLists.txt:
```diff
@@ -11,8 +11,9 @@ add_subdirectory(legacy)
 
 add_subdirectory(extension)
 
+add_subdirectory(image_processing)
```
Given that the `image_processing` tests are now in `test/core/image_processing`, shouldn't the path above reflect that ? (Or rather, that entire line should be added to `test/core/CMakeLists.txt`.)

mloskot replied:
>  (Or rather, that entire line should be added to test/core/CMakeLists.txt.)

Yes. I'd stick it at the end, as line 40:

https://github.com/BoostGSoC19/gil-olzhas/blob/4ed7701b47764edf327e7c82f65cf9815026bdc2/test/core/CMakeLists.txt#L30-L39

simmplecoder committed:
Add math and lexical_cast to get-boost

As math and lexical_cast are used
in lanczos scaling, both were added
to get-boost

simmplecoder committed:
Revert "Add newline at the end of files"

Since Boost.Math.Constants is
overengineered for use case by
requiring Boost.lexical_cast or
a global define, it is removed

This reverts commit 0743ab072ff3455421853697f7f46aee7d22382c.

simmplecoder committed:
Define pi and use in lanczos

Value of pi is moved out into detail
and used by lanczos

simmplecoder committed:
Move image_processing to test/core

simmplecoder committed:
formatting fixes

simmplecoder committed:
Adjust CMakeLists for moved IP tests

This commit removes image_processing as
subdirectory from test/ and adds to
test/core's CMakeLists

simmplecoder committed:
Remove unused from get-boost

Since lanczos scaling no longer uses
boost.math.constants, lexical_cast
has been removed as well
